lib/tbot/tbot.go

```diff
@@ -118,109 +108,120 @@ func (b *Bot) Run(ctx context.Context) error {
+		cfg:               b.cfg,
+		reloadBroadcaster: reloadBroadcaster,
+		log: b.log.WithField(
+			trace.Component, teleport.Component(componentTBot, "identity"),
+		),
+	}
+	// Initialize bot's own identity. This will load from disk, or fetch a new
+	// identity, and perform an initial renewal if necessary.
+	if err := b.botIdentitySvc.Initialize(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	services = append(services, b.botIdentitySvc)
+
+	// Setup all other services
 	if b.cfg.DiagAddr != "" {
-		eg.Go(func() error {
-			b.log.WithField("addr", b.cfg.DiagAddr).Info(
-				"diag_addr configured, diagnostics service will be started.",
-			)
-			mux := http.NewServeMux()
-			mux.Handle("/metrics", promhttp.Handler())
-			// Only expose pprof when `-d` is provided.
-			if b.cfg.Debug {
-				b.log.Info("debug mode enabled, profiling endpoints will be served on the diagnostics service.")
-				mux.HandleFunc("/debug/pprof/", pprof.Index)
-				mux.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
-				mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
-				mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
-				mux.HandleFunc("/debug/pprof/trace", pprof.Trace)
-			}
-			mux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				w.WriteHeader(http.StatusNotFound)
-				msg := "404 - Not Found\n\nI'm a little tbot,\nshort and stout,\nthe page you seek,\nis not about."
-				_, _ = w.Write([]byte(msg))
-			}))
-			srv := http.Server{
-				Addr:              b.cfg.DiagAddr,
-				Handler:           mux,
-				ReadTimeout:       apidefaults.DefaultIOTimeout,
-				ReadHeaderTimeout: defaults.ReadHeadersTimeout,
-				WriteTimeout:      apidefaults.DefaultIOTimeout,
-				IdleTimeout:       apidefaults.DefaultIdleTimeout,
+		services = append(services, &diagnosticsService{
+			diagAddr:     b.cfg.DiagAddr,
+			pprofEnabled: b.cfg.Debug,
+			log: b.log.WithField(
+				trace.Component, teleport.Component(componentTBot, "diagnostics"),
+			),
+		})
+	}
+	services = append(services, &outputsService{
+		botIdentitySrc: b,
+		cfg:            b.cfg,
+		log: b.log.WithField(
+			trace.Component, teleport.Component(componentTBot, "outputs"),
+		),
+		reloadBroadcaster: reloadBroadcaster,
+	})
+	services = append(services, &caRotationService{
+		botIdentitySrc: b,
+		cfg:            b.cfg,
+		log: b.log.WithField(
+			trace.Component, teleport.Component(componentTBot, "ca-rotation"),
+		),
+		reloadBroadcaster: reloadBroadcaster,
+	})
+	// Append any services configured by the user
+	services = append(services, b.cfg.Services...)
+
+	b.log.Info("Initialization complete. Starting services.")
+	// Start services
+	for _, svc := range services {
+		svc := svc
+		log := b.log.WithField("service", svc.String())
+
+		if b.cfg.Oneshot {
+			svc, ok := svc.(bot.OneShotService)
+			// We ignore services with no one-shot implementation
+			if !ok {
+				log.Debug("Service does not support oneshot mode, ignoring.")
```

Comment:
Yeah - I'm not 100% on how I want to do this - but I'm probably going to make those hard-fail and error when they are introduced. It might be I just implement `oneshot` and make it return an error. I figured that it's hard to tell how it'll look until I have a service that behaves that way so I left it like this for now 😓

Reply:
Ah, interesting, that'd be even better! But I agree, we can cross that bridge later.